src/flatten.jl

```diff
@@ -21,20 +21,20 @@ flatten(x) = flatten(Float64, x)
 
 function flatten(::Type{T}, x::Integer) where {T<:Real}
     v = T[]
-    unflatten_to_Integer(v::Vector{T}) = x
+    unflatten_to_Integer(v::AbstractVector{<:Real}) = x
     return v, unflatten_to_Integer
 end
 
 function flatten(::Type{T}, x::R) where {T<:Real,R<:Real}
     v = T[x]
-    unflatten_to_Real(v::Vector{T}) = convert(R, only(v))
+    unflatten_to_Real(v::AbstractVector{<:Real}) = only(v)
```

Comment:
Note that this line will change the current behavior for unflatten quite drastically:
```julia
x = (a = 1., b = [2., 3.], c = [4 5 ; 6 7])
typeof(x.b) #Vector{Float64}
xvec, unflat = flatten(Float16, x)
x2 = unflat(xvec)
typeof(x2.b) #Vector{Float16}
```
I don't think there is any other way though to facilitate AD while keeping initial parameter types.

Reply:
Agreed -- very breaking. From my perspective in terms of how I tend to use ParameterHandling in practice, the new behaviour is more helpful anyway though 🤷

Reply:
I think there is no way around it if you want to work with AD here. Type changes could be a problem if you define a concrete container to collect samples (e.g. MCMC) of your model parameter.